### PR TITLE
Search by shape

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

File: libs/bragi/src/api.rs [hunk 1]
```diff
@@ -39,6 +39,12 @@ use super::query;
 use model::v1::*;
 use model;
 
+static MAX_LAT: f64 = 180f64;
+static MIN_LAT: f64 = -180f64;
+
+static MAX_LON: f64 = 90f64;
+static MIN_LON: f64 = -90f64;
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: better to use const instead of static. That's almost the same thing.

File: libs/bragi/src/api.rs [hunk 2]
```diff
@@ -110,47 +116,80 @@ fn status(&self) -> rustless::Api {
...
+                endpoint.handle(move |client, params| {
+                    let q = match params.find("q") {
+                        Some(val) => val.as_string().unwrap_or("").to_string(),
+                        None => "".to_string(),
+                    };
+                    let goe = params.find_path(&["coordinates"]).unwrap().as_array().unwrap();
+                    let mut shape = Vec::new();
+                    for ar in goe[0].as_array().unwrap() {
+                        shape.push((ar[0].as_f64().unwrap(), ar[1].as_f64().unwrap()));
+                    }
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: That's a lot of unwrap...

File: libs/bragi/src/api.rs [hunk 3]
```diff
@@ -110,47 +116,80 @@ fn status(&self) -> rustless::Api {
...
+                                             detail: Some("you need to provide a lon AND a lat \
+                                                           if you provide one of them"
+                                                 .to_string()),
+                                         })])
+                            } else {
+                                Ok(())
+                            }
                         } else {
-                            Ok(())
+                            panic!("should never happen, already checked");
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: `unreachable!()` is made for this exact pattern.

File: libs/bragi/src/api.rs [hunk 4]
```diff
@@ -110,47 +116,80 @@ fn status(&self) -> rustless::Api {
...
-                        panic!("should never happen, already checked");
-                    }
+                    });
                 });
-            });
-            let cnx = self.es_cnx_string.clone();
-            endpoint.handle(move |client, params| {
+                let cnx = self.es_cnx_string.clone();
+                endpoint.handle(move |client, params| {
                 let q = params.find("q").unwrap().as_string().unwrap().to_string();
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: bad indent

File: libs/bragi/src/query.rs [hunk 5]
```diff
@@ -62,7 +61,7 @@ fn make_place(doc_type: String, value: Option<Box<serde_json::Value>>) -> Option
     })
 }
 
-fn query(q: &String, cnx: &String, match_type: &str) -> Result<Vec<mimir::Place>, rs_es::error::EsError> {
+fn query(q: &String, cnx: &String, match_type: &str, shape: Option<Vec<rs_es::units::Location>>) -> Result<Vec<mimir::Place>, rs_es::error::EsError> {
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: long line

File: libs/bragi/src/query.rs [hunk 6]
```diff
@@ -80,21 +79,24 @@ fn query(q: &String, cnx: &String, match_type: &str) -> Result<Vec<mimir::Place>
 			                                                          .build())
                               .build()])
                         .build();
+    let mut must = vec![rs_q::build_match(match_type, q.to_string())
+             .with_minimum_should_match(rs_es::query::MinimumShouldMatch::from(100f64)).build()];
+
+	if let Some(s) = shape {
+		must.push(rs_q::build_geo_polygon("coord", s).build());
+	}
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: Bad indent. rust fmt may be needed.

File: libs/bragi/src/query.rs [hunk 7]
```diff
@@ -119,26 +121,30 @@ fn query_location(_q: &String,
...
     	//First search with match = "name.prefix".
     	//If no result then another search with match = "name.ngram"
-    	let results = try!(query_prefix(&q, cnx));
+    	fn clone_shape(shape: &Option<Vec<(f64, f64)>>) -> Option<Vec<rs_es::units::Location>> {
+    		shape.as_ref().map(|v| v.iter().map(|&l| l.into()).collect())
+    	}
+    	let results = try!(query_prefix(&q, cnx, clone_shape(&shape)));
         if results.is_empty() {
-        	query_ngram(&q, cnx)
+        	query_ngram(&q, cnx, clone_shape(&shape))
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: `make_shape` is a better name as we don't clone now.

File: libs/bragi/src/api.rs (hunk 1 above)
Comment (azime) [inline libs/bragi/src/api.rs]: ok

File: libs/bragi/src/api.rs (hunk 3 above)
Comment (azime) [inline libs/bragi/src/api.rs]: Ok

File: libs/bragi/src/api.rs [hunk 8]
```diff
@@ -49,108 +55,148 @@ fn render<T>(mut client: rustless::Client,
...
+                        });
+                    })
+                });
+
+                let cnx = self.es_cnx_string.clone();
+                endpoint.handle(move |client, params| {
+                    let q = match params.find("q") {
+                        Some(val) => val.as_string().unwrap_or("").to_string(),
+                        None => "".to_string(),
+                    };
```
Comment (TeXitoi) [inline libs/bragi/src/api.rs]: Really minor, but you can

```rust
let q = match params.find("q") {
    Some(val) => val.as_string().unwrap_or(""),
    None => "",
}.to_string();
```

File: libs/bragi/src/query.rs [hunk 9]
```diff
@@ -62,7 +61,11 @@ fn make_place(doc_type: String, value: Option<Box<serde_json::Value>>) -> Option
     })
 }
 
-fn query(q: &String, cnx: &String, match_type: &str) -> Result<Vec<mimir::Place>, rs_es::error::EsError> {
+fn query(q: &String,
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: `q` should be a `&str`

File: libs/bragi/src/query.rs [hunk 10]
```diff
@@ -62,7 +61,11 @@ fn make_place(doc_type: String, value: Option<Box<serde_json::Value>>) -> Option
...
+         cnx: &String,
```
Comment (TeXitoi) [inline libs/bragi/src/query.rs]: same

File: libs/bragi/src/api.rs (hunk 8 above)
Comment (azime) [inline libs/bragi/src/api.rs]: ok

File: libs/bragi/src/api.rs (hunk 4 above)
Comment (azime) [inline libs/bragi/src/api.rs]: ok

File: libs/bragi/src/query.rs (hunk 9 above)
Comment (azime) [inline libs/bragi/src/query.rs]: ok

File: libs/bragi/src/query.rs (hunk 10 above)
Comment (azime) [inline libs/bragi/src/query.rs]: ok

File: libs/bragi/src/query.rs (hunk 6 above)
Comment (azime) [inline libs/bragi/src/query.rs]: ok

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**